src/agent/agent.c

```diff
@@ -30,6 +31,65 @@ static bool
                                 void *userdata,
                                 free_func_t free_userdata);
 
+#define AGENT_HEARTBEAT_INTERVAL_USEC (1000000)
+
+static int agent_reset_heartbeat_timer(Agent *agent, sd_event_source **event_source);
+
+static int agent_heartbeat_timer_callback(
+                        sd_event_source *event_source,
+                        UNUSED uint64_t usec,
+                        void *userdata) {
+        Agent *agent = userdata;
+        int r;
+
+        assert(event_source);
+
+        static uint32_t count;
+        printf("%s(): count = %d\n", __func__, count++);
+
+        r = sd_bus_emit_signal(
+                    agent->peer_dbus,
+                    INTERNAL_MANAGER_OBJECT_PATH,
+                    INTERNAL_MANAGER_INTERFACE,
+                    "Heartbeat",
+                    "s",
+                    agent->name);
+        if (r < 0) {
+                fprintf(stderr, "Failed to emit heartbeat signal: %m");
+                return r;
+        }
+
+        r = agent_reset_heartbeat_timer(agent, &event_source);
+        if (r < 0) {
+                fprintf(stdout, "Failed to reset periodic timer event source, ignoring: %m");
+                return r;
+        }
+
+        return 0;
+}
+
+static int agent_reset_heartbeat_timer(
+                        Agent *agent,
+                        sd_event_source **event_source) {
+        sd_event *event = agent->event;
+        return event_reset_time_relative(event, event_source, CLOCK_BOOTTIME, AGENT_HEARTBEAT_INTERVAL_USEC, 0,
+                                         agent_heartbeat_timer_callback, agent, 0, "periodic-timer-event-source", false);
+}
+
+static int agent_setup_heartbeat_timer(Agent *agent) {
+        _cleanup_(sd_event_source_unrefp) sd_event_source *event_source = NULL;
+        int r;
+
+        assert(agent);
+
+        r = agent_reset_heartbeat_timer(agent, &event_source);
+        if (r < 0) {
+                fprintf(stdout, "Failed to reset heartbeat timer: %m");
+                return r;
+        }
+
+        return sd_event_source_set_floating(event_source, true);
```

Comment:
I don't see the need for a floating reference here, that just makes things complex. Why not just store an owning reference to the event in the agent.

Reply:
Currently the event source object is allocated in ```sd_event_add_time()```, called in ```event_reset_time()```, and ```event_reset_time()``` decides whether to call ```sd_event_add_time()``` based on whether the event loop object has been allocated yet.  I kind of like the floating mode being used here, where the event loop tracks the lifetime of the event source object.   I think it makes sense in this case because the event source object is only used for one purpose, for a timer, and there isn't a need for the application to touch it again.